django/db/backends/postgresql/base.py

```diff
@@ -1,67 +1,111 @@
+    raise ImproperlyConfigured("Error loading psycopg module: %s" % e)
 
 
-def psycopg2_version():
-    version = psycopg2.__version__.split(" ", 1)[0]
+def psycopg_version():
+    version = Database.__version__.split(" ", 1)[0]
     return get_version_tuple(version)
 
 
-PSYCOPG2_VERSION = psycopg2_version()
+PSYCOPG_VERSION = psycopg_version()
 
-if PSYCOPG2_VERSION < (2, 8, 4):
+
+if PSYCOPG_VERSION < (2, 8, 4):
     raise ImproperlyConfigured(
         "psycopg2 version 2.8.4 or newer is required; you have %s"
-        % psycopg2.__version__
+        % Database.__version__
     )
 
+if PSYCOPG_VERSION[0] >= 3:
+    import psycopg
+    from psycopg import sql
+    from psycopg.pq import Format
+    from psycopg.types.datetime import TimestamptzLoader
+    from psycopg.types.range import Range, RangeDumper
+    from psycopg.types.string import StrDumper, TextLoader
+
+    TIMESTAMPTZ_OID = psycopg.adapters.types["timestamptz"].oid
+    TSRANGE_OID = psycopg.postgres.types["tsrange"].oid
+    TSTZRANGE_OID = psycopg.postgres.types["tstzrange"].oid
+else:
+    import psycopg2.extensions
+    import psycopg2.extras
+
+    psycopg2.extensions.register_adapter(SafeString, psycopg2.extensions.QuotedString)
+    psycopg2.extras.register_uuid()
+
+    # Register support for inet[] manually so we don't have to handle the Inet()
+    # object on load all the time.
+    INETARRAY_OID = 1041
+    INETARRAY = psycopg2.extensions.new_array_type(
+        (INETARRAY_OID,),
+        "INETARRAY",
+        psycopg2.extensions.UNICODE,
+    )
+    psycopg2.extensions.register_type(INETARRAY)
+
+# Some of these import psycopg, so import them after checking if it's installed.
+from .client import DatabaseClient  # NOQA isort:skip
+from .creation import DatabaseCreation  # NOQA isort:skip
+from .features import DatabaseFeatures  # NOQA isort:skip
+from .introspection import DatabaseIntrospection  # NOQA isort:skip
+from .operations import DatabaseOperations  # NOQA isort:skip
+from .psycopg_any import IsolationLevel  # NOQA isort:skip
+from .schema import DatabaseSchemaEditor  # NOQA isort:skip
+
 
-# Some of these import psycopg2, so import them after checking if it's installed.
-from .client import DatabaseClient  # NOQA
-from .creation import DatabaseCreation  # NOQA
-from .features import DatabaseFeatures  # NOQA
-from .introspection import DatabaseIntrospection  # NOQA
-from .operations import DatabaseOperations  # NOQA
-from .schema import DatabaseSchemaEditor  # NOQA
+@lru_cache
+def get_adapters_template(use_tz, timezone):
+    # Create at adapters map extending the base one to base connections on
+    ctx = psycopg.adapt.AdaptersMap(psycopg.adapters)
 
-psycopg2.extensions.register_adapter(SafeString, psycopg2.extensions.QuotedString)
-psycopg2.extras.register_uuid()
+    # Register a no-op dumper to avoid a round trip from psycopg3's decode
+    # to json.dumps() to json.loads(), when using a custom decoder in
+    # JSONField.
```

Comment:
```suggestion
    # Register a no-op dumper to avoid a round trip from psycopg version 3
    # decode to json.dumps() to json.loads(), when using a custom decoder
    # in JSONField.
```